File: python-client/jobicy.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from html import unescape
from math import isfinite
import re
from typing import Any
from urllib.parse import urlsplit

import requests


class JobicyError(Exception):
    pass


class JobicyRateLimitError(JobicyError):
    def __init__(self, message: str, retry_after_seconds: int | None = None) -> None:
        super().__init__(message)
        self.retry_after_seconds = retry_after_seconds


@dataclass(frozen=True, slots=True)
class Job:
    id: int | str
    url: str
    job_title: str
    company_name: str
    company_logo: str | None
    industries: tuple[str, ...]
    job_types: tuple[str, ...]
    location: str
    level: str | None
    excerpt: str
    published_at: str | None
    salary_min: float | None
    salary_max: float | None
    salary_currency: str | None
    salary_period: str | None

    @classmethod
    def from_api(cls, payload: dict[str, Any]) -> Job:
        def text(key: str, default: str = "") -> str:
            value = payload.get(key)
            return unescape(str(value)).strip() if value is not None else default

        def amount(key: str) -> float | None:
            value = payload.get(key)

            if value in (None, "") or isinstance(value, bool):
                return None

            try:
                number = float(value)
            except (TypeError, ValueError):
                return None

            return number if isfinite(number) and number > 0 else None

        def labels(key: str) -> tuple[str, ...]:
            values = payload.get(key)
            if not isinstance(values, list):
                return ()
            return tuple(unescape(str(value)).strip() for value in values if str(value).strip())

        return cls(
            id=payload["id"],
            url=str(payload["url"]),
            job_title=text("jobTitle") or "Remote opportunity",
            company_name=text("companyName") or "Company not specified",
            company_logo=text("companyLogo") or None,
            industries=labels("jobIndustry"),
            job_types=labels("jobType"),
            location=text("jobGeo") or "Location not specified",
            level=text("jobLevel") or None,
            excerpt=re.sub(r"\s+", " ", re.sub(r"<[^>]*>", " ", text("jobExcerpt"))).strip(),
            published_at=text("pubDate") or None,
            salary_min=amount("salaryMin"),
            salary_max=amount("salaryMax"),
            salary_currency=text("salaryCurrency") or None,
            salary_period=text("salaryPeriod") or None,
        )

    @property
    def salary_display(self) -> str:
        if self.salary_min is None and self.salary_max is None:
            return ""

        currency = self.salary_currency or "USD"

        if self.salary_min is not None and self.salary_max is not None and self.salary_min != self.salary_max:
            amount = f"{self.salary_min:,.0f}–{self.salary_max:,.0f}"
        else:
            amount = f"{self.salary_min if self.salary_min is not None else self.salary_max:,.0f}"

        period = f" / {self.salary_period}" if self.salary_period else ""
        return f"{currency} {amount}{period}"

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


class JobicyClient:
    api_url = "https://jobicy.com/api/v2/remote-jobs"

    def __init__(self, timeout: float = 15.0, session: requests.Session | None = None) -> None:
        if not isinstance(timeout, (int, float)) or timeout <= 0:
            raise ValueError("timeout must be a positive number of seconds")

        self.timeout = float(timeout)
        self.session = session or requests.Session()
        self.session.headers.update({
            "Accept": "application/json",
            "User-Agent": "Jobicy-Integration-Example/python-client",
        })

    def get_jobs(
        self,
        count: int = 50,
        geo: str | None = None,
        industry: str | None = None,
        tag: str | None = None,
    ) -> list[Job]:
        if isinstance(count, bool) or not isinstance(count, int) or not 1 <= count <= 100:
            raise ValueError("count must be an integer between 1 and 100")

        parameters: dict[str, str | int] = {"count": count}

        for name, value in (("geo", geo), ("industry", industry), ("tag", tag)):
            if value is None:
                continue
            if not isinstance(value, str):
                raise ValueError(f"{name} must be a string")
            if value.strip():
                parameters[name] = value.strip()

        try:
            response = self.session.get(self.api_url, params=parameters, timeout=self.timeout)
        except requests.Timeout as error:
            raise JobicyError(f"Jobicy request timed out after {self.timeout:g} seconds") from error
        except requests.RequestException as error:
            raise JobicyError(f"Jobicy request failed: {error}") from error

        if response.status_code == 429:
            retry_header = response.headers.get("Retry-After", "")
            retry_after = int(retry_header) if retry_header.isdigit() else None
            raise JobicyRateLimitError("Jobicy temporarily rate limited the request", retry_after)

        if not response.ok:
            raise JobicyError(f"Jobicy API returned HTTP {response.status_code}")

        try:
            payload = response.json()
        except (ValueError, requests.RequestException) as error:
            raise JobicyError("Jobicy API returned invalid JSON") from error

        if not isinstance(payload, dict) or not isinstance(payload.get("jobs"), list):
            raise JobicyError("Jobicy API response does not contain a jobs array")

        unique: dict[str, Job] = {}

        for item in payload["jobs"]:
            if not isinstance(item, dict) or item.get("id") is None or not isinstance(item.get("url"), str):
                continue

            parsed_url = urlsplit(item["url"])
            if parsed_url.scheme != "https" or parsed_url.hostname != "jobicy.com":
                continue

            try:
                job = Job.from_api(item)
            except (TypeError, ValueError, KeyError):
                continue

            unique[str(job.id)] = job

        return list(unique.values())

    def close(self) -> None:
        self.session.close()

    def __enter__(self) -> JobicyClient:
        return self

    def __exit__(self, exc_type: object, exc_value: object, traceback: object) -> None:
        self.close()

File: telegram-bot/src/jobicy.js
const API_URL = "https://jobicy.com/api/v2/remote-jobs";
const MAX_COUNT = 100;
const REQUEST_TIMEOUT_MS = 15_000;

export class JobicyRequestError extends Error {
  constructor(message, status = null, retryAfterSeconds = null) {
    super(message);
    this.name = "JobicyRequestError";
    this.status = status;
    this.retryAfterSeconds = retryAfterSeconds;
  }
}

export async function fetchJobs({ geo = "", industry = "", keywords = "", count = 50, project }) {
  const limit = Math.max(1, Math.min(MAX_COUNT, Number.parseInt(String(count), 10) || 50));
  const url = new URL(API_URL);
  url.searchParams.set("count", String(limit));

  if (geo.trim()) url.searchParams.set("geo", geo.trim());
  if (industry.trim()) url.searchParams.set("industry", industry.trim());

  const terms = keywords.split(",").map((value) => value.trim().toLowerCase()).filter(Boolean);

  if (terms.length === 1) url.searchParams.set("tag", terms[0]);

  let response;

  try {
    response = await fetch(url, {
      headers: {
        Accept: "application/json",
        "User-Agent": `Jobicy-Integration-Example/${project}`
      },
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS)
    });
  } catch (error) {
    throw new JobicyRequestError(`Jobicy request failed: ${error.message}`);
  }

  if (!response.ok) {
    const retryHeader = response.headers.get("retry-after");
    const retryAfterSeconds = retryHeader && /^\d+$/.test(retryHeader) ? Number(retryHeader) : null;
    throw new JobicyRequestError(`Jobicy API returned HTTP ${response.status}`, response.status, retryAfterSeconds);
  }

  let payload;

  try {
    payload = await response.json();
  } catch {
    throw new JobicyRequestError("Jobicy API returned invalid JSON");
  }

  if (!payload || !Array.isArray(payload.jobs)) {
    throw new JobicyRequestError("Jobicy API response does not contain a jobs array");
  }

  const unique = new Map();

  for (const job of payload.jobs) {
    if (!job || typeof job !== "object" || job.id == null || typeof job.url !== "string") continue;

    let canonical;

    try {
      canonical = new URL(job.url);
    } catch {
      continue;
    }

    if (canonical.protocol !== "https:" || canonical.hostname !== "jobicy.com") continue;

    const searchable = [
      job.jobTitle,
      job.companyName,
      job.jobGeo,
      job.jobExcerpt,
      ...(Array.isArray(job.jobIndustry) ? job.jobIndustry : []),
      ...(Array.isArray(job.jobType) ? job.jobType : [])
    ].filter(Boolean).join(" ").toLowerCase();

    if (terms.length && !terms.some((term) => searchable.includes(term))) continue;

    unique.set(String(job.id), job);
  }

  return [...unique.values()];
}

export function formatSalary(job) {
  const minimum = Number(job.salaryMin);
  const maximum = Number(job.salaryMax);
  const hasMinimum = Number.isFinite(minimum) && minimum > 0;
  const hasMaximum = Number.isFinite(maximum) && maximum > 0;

  if (!hasMinimum && !hasMaximum) return "";

  const currency = typeof job.salaryCurrency === "string" && /^[A-Z]{3}$/i.test(job.salaryCurrency)
    ? job.salaryCurrency.toUpperCase()
    : "USD";
  const formatter = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency,
    maximumFractionDigits: 0
  });
  const amount = hasMinimum && hasMaximum && minimum !== maximum
    ? `${formatter.format(minimum)}–${formatter.format(maximum)}`
    : formatter.format(hasMinimum ? minimum : maximum);
  const period = typeof job.salaryPeriod === "string" && job.salaryPeriod.trim()
    ? ` / ${job.salaryPeriod.trim()}`
    : "";

  return `${amount}${period}`;
}

export function normalizeText(value, maxLength = 300) {
  const text = String(value ?? "")
    .replace(/<[^>]*>/g, " ")
    .replace(/&(?:amp|lt|gt|quot|apos|#39|nbsp|hellip);/gi, (entity) => ({
      "&amp;": "&",
      "&lt;": "<",
      "&gt;": ">",
      "&quot;": '"',
      "&apos;": "'",
      "&#39;": "'",
      "&nbsp;": " ",
      "&hellip;": "…"
    }[entity.toLowerCase()] || entity))
    .replace(/\s+/g, " ")
    .trim();

  return text.length > maxLength ? `${text.slice(0, maxLength - 1)}…` : text;
}
